File: rob599_basic/src/service.py
#!/usr/bin/env python3

# Example of a simple service.
#
# service.py
#
# Bill Smart
#
# This shows how to put together a simple service client in Python.  This depends on the service message
# being defined and compiled first.


# Import what we need.
import rospy
import sys


# Import the service message definitions that we need, from the current package.  Note that we use
# rob599_basic.srv as the package name.  For the server, we need the base message (Doubler) and the
# response (DoublerResponse), which is automatically created for us by catkin.
from rob599_basic.srv import Doubler, DoublerResponse


# This callback is called whenever a service request comes in.  The parameter is of type DoublerRequest, and
# the return, of type DoublerResponse, is returned by the service call mechanism.
def callback(request):
	rospy.loginfo('Got {0}'.format(request.number))

	# The return from the callback is used as the response from the servic call.
	return DoublerResponse(request.number * 2)


if __name__ == '__main__':
	# Initialize the node.
	rospy.init_node('server')

	# Create the service, specifying the name, the service message base type, and the callback function
	# to be used to fulfil the requests.
	service = rospy.Service('doubler', Doubler, callback)

	rospy.loginfo('Service started')

	# Give control over to ROS.
	rospy.spin()

File: rob599_basic/src/service.cpp
// A basic service server in C++
//
// service.cpp
//
// Bill Smart
//
// This shows how to set up a basic service service in C++


// Include the basic ROS stuff.
#include <ros/ros.h>

// Include the definition of the Doubler message.  This was created for us by catkin_make, and lives in the
// catkin workspace.
#include <rob599_basic/Doubler.h>


// This is the callback for the service.  It takes two arguments, one of the Request type, and one of the
// Response type.  Both of them are (mutable) references.  Unlike Python, the service callback does not
// return the Response.  Instead, set the member variables of the response, and return true.
bool doubler(rob599_basic::Doubler::Request &request, rob599_basic::Doubler::Response &response) {
	// Send a message to the info log channel.
	ROS_INFO("Got %d", request.number);

	// Store the response value in the response argument.
	response.doubled = request.number * 2;

	return true;
}


int main(int argc, char **argv) {
	// Initialize the node and make a node handle.
	ros::init(argc, argv, "server");
	ros::NodeHandle node;

	// Set up the service, using NodeHandle.advertiseService.  This takes a service name, and a callback
	// function.
	ros::ServiceServer service = node.advertiseService("doubler", doubler);
	ROS_INFO("Service started");

	// Give control over to ROS.
	ros::spin();

	return 0;
}